bothan-core/src/store.rs

```diff
@@ -1,27 +1,120 @@
+use std::collections::{HashMap, HashSet};
+
+use tokio::sync::{Mutex, RwLock};
+
 use crate::types::AssetInfo;
 
-pub enum AssetStatus {
-    Unsupported,
-    Pending,
-    Available(AssetInfo),
+// TODO: Store should have namespace for different sources
+pub struct Store {
+    asset_store: RwLock<HashMap<String, AssetInfo>>,
+    query_ids: Mutex<HashSet<String>>,
 }
 
-#[derive(thiserror::Error, Debug)]
-pub enum Error {
-    #[error("Not started")]
-    NotStarted,
-
-    #[error("failed to modify query IDs: {0}")]
-    ModifyQueryIDsFailed(String),
+impl Default for Store {
+    fn default() -> Self {
+        Store::new()
+    }
 }
 
-/// Type alias for a service result, which is either a valid result or an error.
-/// The universal trait for all services that provide price data.
-#[async_trait::async_trait]
-pub trait AssetStore: Send + Sync + 'static {
-    async fn start(&mut self);
-    async fn get_prices(&self, ids: &[&str]) -> Vec<AssetStatus>;
-    async fn add_query_ids(&mut self, ids: &[&str]) -> Result<(), Error>;
-    async fn remove_query_ids(&mut self, ids: &[&str]) -> Result<(), Error>;
-    async fn get_query_ids(&self) -> Vec<String>;
+impl Store {
+    pub fn new() -> Self {
+        Self {
+            asset_store: RwLock::new(HashMap::new()),
+            query_ids: Mutex::new(HashSet::new()),
+        }
+    }
+
+    pub async fn get_assets<K: AsRef<str> + Send + Sync>(
+        &self,
+        ids: &[K],
+    ) -> Vec<crate::worker::AssetStatus> {
+        let data_store = self.asset_store.read().await;
+        let query_ids = self.query_ids.lock().await;
+
+        ids.iter()
+            .map(
+                |id| match (query_ids.contains(id.as_ref()), data_store.get(id.as_ref())) {
+                    (false, _) => crate::worker::AssetStatus::Unsupported,
+                    (true, Some(asset)) => crate::worker::AssetStatus::Available(asset.clone()),
+                    (true, None) => crate::worker::AssetStatus::Pending,
+                },
+            )
+            .collect()
+    }
+
+    pub async fn get_all_assets(&self) -> Vec<AssetInfo> {
+        self.asset_store.read().await.values().cloned().collect()
+    }
+
+    pub async fn set_asset<K: Into<String>>(&self, id: K, asset_info: AssetInfo) {
+        let mut data_store = self.asset_store.write().await;
+        data_store.insert(id.into(), asset_info);
+    }
+
+    pub async fn set_assets<K: Into<String>>(&self, assets: Vec<(K, AssetInfo)>) {
+        let mut data_store = self.asset_store.write().await;
+        for (id, asset) in assets {
+            data_store.insert(id.into(), asset);
+        }
+    }
+
+    pub async fn add_query_ids<K: Into<String>>(&self, ids: Vec<K>) -> Vec<String> {
```

Comment:
K is clear for store since its the key value kept in the store but T is generic enough to be used for get_assets() since we're not exactly treating it as a specific key value

Reply:
I think change`AssetWorker` template on `K` should be ok